ferminet/networks.py

```diff
@@ -440,17 +440,30 @@ def logdet_matmul(xs: Sequence[jnp.ndarray],
     determinant (or product of the i-th determinant in each spin channel, if
     full_det is not used).
   """
-  slogdets = [slogdet(x) for x in xs]
-  sign_in, logdet = functools.reduce(
-      lambda a, b: (a[0]*b[0], a[1]+b[1]), slogdets)
+  # Special case if there is only one electron in any channel
+  # We can avoid the log(0) issue by not going into the log domain
+  dets = [x.reshape(-1) for x in xs if x.shape[-1] == 1]
+  dets = functools.reduce(
+    lambda a, b: a*b, dets
+  ) if len(dets) > 0 else 1
```

Comment:
The guard that dets is not empty is not required -- this can only happen if we have no electrons and an exception is raised (albeit with an unhelpful error message) much earlier in that instance.

Reply:
`len(dets)` may be empty if all channels have at least 2 electrons. In that case, we want to set the `dets` variable to 1 (so it is neglected in the multiplication in line 458.

I think you may have missed the condition in line 445. There may be electrons and `dets` be empty.

Reply:
Sorry, this had slipped off my radar. I had indeed missed the condition. Having said that, I think there's a simpler, clearer way.

Looking at this with fresh eyes, it occurred to me that we can simplify this logic by passing the initial value to `functools.reduce`:

```python
  det1 = functools.reduce(lambda a, b: a * b,                                                                                                            
                          [x.reshape(-1) for x in xs if x.shape[-1] == 1], 1)                                                                            
  sign_in, logdet = functools.reduce(                                                                                                                   
      lambda a, b: (a[0] * b[0], a[1] + b[1]),                                                                                                          
      [slogdet(x) for x in xs if x.shape[-1] > 1], (1, 1))
  # log-sum-exp trick
  maxlogdet = jnp.max(logdet)
  det = sign_in * det1 *jnp.exp(logdet - maxlogdet)
```

(with, of course, an appropriate comment describing why we consider 1x1 matrices separately). This ensures `det1`, `sign_in` and `logdet` are all 1 if the appropriate list is empty.

Maybe nicer is to use [`jax.tree_reduce`](https://jax.readthedocs.io/en/latest/_autosummary/jax.tree_util.tree_reduce.html#jax.tree_util.tree_reduce) as (unlike `functools.reduce`) it takes keyword arguments. It's just a lightweight wrapper around `functools.reduce` in this case though.

What do you think?

Reply:
I agree, using the default arguments in `functools.reduce` seems to be a better way to code this!

Reply:
Looks great, thanks!